File: frontend/src/types/contracts/report.contract.ts
/**
 * 报告相关类型契约
 * 与backend/app/schemas/contracts/report_contract.py完全对应
 */

export enum ReportFormat {
  FULL = "full",
  SUMMARY = "summary",
  INSIGHTS = "insights"
}

export interface InsightItem {
  title: string;
  content: string;
  confidence: number; // 0.0 - 1.0
  source_count: number;
  tags: string[];
}

// 5个关键字段的类型定义
export interface ExecutiveSummary {
  headline?: string;
  total_communities: number;
  key_insights: number;
  top_opportunity?: string;
  confidence_score?: number;
  summary_points: string[];
}

export interface MarketMetrics {
  total_mentions: number;
  sentiment_score: number;
  top_communities: string[];
  trending_keywords: string[];
  engagement_rate?: number;
  sample_size?: number;
}

export interface PainPointExample {
  post_id: string;
  community?: string;
  permalink?: string;
  content_snippet?: string;
  upvotes?: number;
}

export interface PainPointInsight {
  description: string;
  sentiment_score: number;
  frequency: number;
  confidence?: number;
  severity?: "low" | "medium" | "high";
  categories: string[];
  example_posts: PainPointExample[];
  tags: string[];
}

export interface CompetitorInsight {
  name: string;
  description?: string;
  market_position?: "leader" | "challenger" | "follower" | "niche";
  mention_count: number;
  sentiment_score: number;
  strengths: string[];
  weaknesses: string[];
  market_share_estimate?: number;
}

export interface OpportunityInsight {
  title: string;
  description: string;
  potential: "low" | "medium" | "high";
  difficulty: "easy" | "medium" | "hard";
  market_size?: string;
  confidence?: number;
  timeframe?: string;
  key_insights: string[];
}

export interface ReportData {
  // 基本元数据
  task_id: string;
  query: string;
  total_posts: number;
  total_comments: number;
  analysis_duration: number;
  confidence_score?: number;

  // 核心洞察
  key_insights: InsightItem[];
  sentiment_summary: Record<string, number>;
  trending_topics: string[];
  user_personas: Array<Record<string, unknown>>;

  // 元数据
  generated_at: string;
  data_freshness: string;
  html_content?: string; // 可选字段

  // 5个关键字段
  executive_summary: ExecutiveSummary;
  market_metrics: MarketMetrics;
  pain_points: PainPointInsight[];
  competitors: CompetitorInsight[];
  opportunities: OpportunityInsight[];

  // 前端扩展的可视化数据
  charts?: ChartData[];
  visualization_config?: VisualizationConfig;
}

// 前端专用的可视化类型
export interface ChartData {
  id: string;
  type: 'pie' | 'bar' | 'line' | 'radar';
  title: string;
  data: Array<{
    label: string;
    value: number;
    color?: string;
  }>;
}

export interface VisualizationConfig {
  theme: 'light' | 'dark';
  colorScheme: string[];
  responsive: boolean;
  animations: boolean;
}

// 报告页面组件Props
export interface ReportPageProps {
  taskId: string;
  format?: ReportFormat;
  onExport?: (format: 'pdf' | 'json') => void;
  onShare?: () => void;
}

// 子组件Props类型
export interface ExecutiveSummaryProps {
  executiveSummary?: ExecutiveSummary;
  totalPosts?: number;
  totalComments?: number;
  confidence?: number;
  sentimentSummary?: Record<string, number>;
}

export interface PainPointsListProps {
  painPoints?: PainPointInsight[];
  sentimentSummary?: Record<string, number>;
  onInsightClick?: (insight: PainPointInsight) => void;
}

export interface CompetitorAnalysisProps {
  competitors?: CompetitorInsight[];
  onCompetitorSelect?: (competitor: CompetitorInsight) => void;
}

export interface OpportunityMatrixProps {
  opportunities?: OpportunityInsight[];
  onOpportunityClick?: (opportunity: OpportunityInsight) => void;
}

File: backend/app/schemas/contracts/report_contract.py
from __future__ import annotations

from enum import Enum
from typing import Dict, List, Optional, Literal

from pydantic import BaseModel, Field

from ...core.types import JsonValue


class ReportFormat(str, Enum):
    FULL = "full"
    SUMMARY = "summary"
    INSIGHTS = "insights"


class InsightItem(BaseModel):
    title: str
    content: str
    confidence: float = Field(ge=0.0, le=1.0)
    source_count: int = Field(ge=0)
    tags: List[str] = Field(default_factory=list)


class ExecutiveSummary(BaseModel):
    headline: Optional[str] = Field(default=None)
    total_communities: int = Field(default=0, ge=0)
    key_insights: int = Field(default=0, ge=0)
    top_opportunity: Optional[str] = Field(default=None)
    confidence_score: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    summary_points: List[str] = Field(default_factory=list)


class MarketMetrics(BaseModel):
    total_mentions: int = Field(default=0, ge=0)
    sentiment_score: float = Field(default=0.0, ge=-1.0, le=1.0)
    top_communities: List[str] = Field(default_factory=list)
    trending_keywords: List[str] = Field(default_factory=list)
    engagement_rate: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    sample_size: Optional[int] = Field(default=None, ge=0)


class PainPointExample(BaseModel):
    post_id: str
    community: Optional[str] = Field(default=None)
    permalink: Optional[str] = Field(default=None)
    content_snippet: Optional[str] = Field(default=None)
    upvotes: Optional[int] = Field(default=None, ge=0)


class PainPointInsight(BaseModel):
    description: str = Field(..., min_length=1)
    sentiment_score: float = Field(..., ge=-1.0, le=1.0)
    frequency: int = Field(..., ge=0)
    confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    severity: Optional[Literal["low", "medium", "high"]] = Field(default=None)
    categories: List[str] = Field(default_factory=list)
    example_posts: List[PainPointExample] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)


class CompetitorInsight(BaseModel):
    name: str = Field(..., min_length=1)
    description: Optional[str] = Field(default=None)
    market_position: Optional[Literal["leader", "challenger", "follower", "niche"]] = Field(default=None)
    mention_count: int = Field(default=0, ge=0)
    sentiment_score: float = Field(default=0.0, ge=-1.0, le=1.0)
    strengths: List[str] = Field(default_factory=list)
    weaknesses: List[str] = Field(default_factory=list)
    market_share_estimate: Optional[float] = Field(default=None, ge=0.0, le=1.0)


class OpportunityInsight(BaseModel):
    title: str = Field(..., min_length=1)
    description: str = Field(..., min_length=1)
    potential: Literal["low", "medium", "high"] = Field(default="medium")
    difficulty: Literal["easy", "medium", "hard"] = Field(default="medium")
    market_size: Optional[str] = Field(default=None)
    confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    timeframe: Optional[str] = Field(default=None)
    key_insights: List[str] = Field(default_factory=list)





class ReportData(BaseModel):
    """标准化报告数据结构（与现有SignalReport兼容的超集）。"""

    # 基本元数据
    task_id: str
    query: str
    total_posts: int = Field(ge=0)
    total_comments: int = Field(ge=0)
    analysis_duration: float = Field(ge=0)
    confidence_score: Optional[float] = Field(default=None, ge=0.0, le=1.0)

    # 核心洞察
    key_insights: List[InsightItem] = Field(default_factory=list)
    sentiment_summary: Dict[str, float] = Field(default_factory=dict)
    trending_topics: List[str] = Field(default_factory=list)
    user_personas: List[Dict[str, JsonValue]] = Field(default_factory=list)

    # 其他
    generated_at: str
    data_freshness: str
    html_content: Optional[str] = None
    data_coverage: Optional[Dict[str, JsonValue]] = None

    # 结构化洞察
    executive_summary: ExecutiveSummary = Field(default_factory=ExecutiveSummary)
    market_metrics: MarketMetrics = Field(default_factory=MarketMetrics)
    pain_points: List[PainPointInsight] = Field(default_factory=list)
    competitors: List[CompetitorInsight] = Field(default_factory=list)
    opportunities: List[OpportunityInsight] = Field(default_factory=list)

    class Config:
        # 兼容服务端额外字段（如 data_coverage 等），避免解析失败
        extra = "ignore"
